### PR TITLE
Feature #128300 feat: TJreports API

### PR DESCRIPTION
**Pull request by monikaatvowel**

### CODE REVIEW

File: src/language/en-GB/en-GB.plg_api_reports.ini [hunk 1]
```diff
@@ -0,0 +1 @@
+PLG_REPORTS_API_REPORT_NAME_MISSSING="Please enter report name"
```
Comment (vaivk369) [inline src/language/en-GB/en-GB.plg_api_reports.ini]: @monikaatvowel  either use  PLG_REPORTS_API this or PLG_API_REPORTS this as a prefix

File: src/language/en-GB/en-GB.plg_api_reports.sys.ini [hunk 2]
```diff
@@ -0,0 +1 @@
+PLG_API_REPORTS = "TJ Reports Api Plugin"
```
Comment (vaivk369) [inline src/language/en-GB/en-GB.plg_api_reports.sys.ini]: @monikaatvowel either use PLG_REPORTS_API this or PLG_API_REPORTS this as a prefix

File: src/reports.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+<extension version="3.0.0" type="plugin" group="api" method="upgrade">
+    <name>PLG_API_REPORTS</name>
+    <version>1.0</version>
+    <creationDate>28 May 2018</creationDate>
+    <author>Techjoomla</author>
+    <authorEmail>extensions@techjoomla.com</authorEmail>
+    <authorUrl>www.techjoomla.com</authorUrl>
+    <copyright>Techjoomla. All rights reserved.</copyright>
+    <license>http://www.gnu.org/licenses/gpl-2.0.html GNU/GPL</license>
+    <description>Tj-Reports API plugin</description>
```
Comment (vaivk369) [inline src/reports.xml]: description should also a be language constant

File: src/reports/filters.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Tjreports API report class
+ *
+ * @since  1.0.0
+ */
+class ReportsApiResourceFilters extends ApiResource
+{
+	/**
+	 * Function save report.
```
Comment (vaivk369) [inline src/reports/filters.php]: @monikaatvowel  this is not the correct function comment

File: src/reports/filters.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	 *
+	 * @return boolean
+	 */
+	public function get()
+	{
+		$user = JFactory::getUser();
```
Comment (vaivk369) [inline src/reports/filters.php]: @monikaatvowel  do you want user to login to site first and then use the API?

File: src/reports/report.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	/**
+	 * Function to get report data from tjreports plugin
+	 *
+	 * @return json
+	 */
+	public function post()
+	{
+		$user = JFactory::getUser();
+
+		if (!$user->id)
```
Comment (vaivk369) [inline src/reports/report.php]: @monikaatvowel Please check other API plugins. Not authorised error are handled by com_api and not by individual plugin

File: src/reports/report.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		}
+
+		$app         = JFactory::getApplication();
+		$jinput      = $app->input;
+		$formData    = $jinput->post;
+		$reportName  = $formData->get('report');
+
+		if (!isset($reportName))
+		{
+			ApiError::raiseError(500, "Please enter report name", 'APIValidationException');
```
Comment (vaivk369) [inline src/reports/report.php]: @monikaatvowel  Lang constant

Comment (vaivk369) [reply src/reports/report.php]: also error code should be 400

File: src/reports/report.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+		$reportFilters  = ($formData->get('filters'))?$formData->get('filters'):[];
+		$reportCols  = ($formData->get('colToshow'))?$formData->get('colToshow'):[];
+
+		JLoader::import('plugins.tjreports.' . $reportName . "." . $reportName, JPATH_SITE);
+		$className = 'TjreportsModel' . ucfirst($reportName);
+
+		if (!class_exists($className))
+		{
+			ApiError::raiseError(500, "Please enter valid report name", 'APIValidationException');
```
Comment (vaivk369) [inline src/reports/report.php]: code 400 and lang constant

File: src/reports/report.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	 * Function to get report data from tjreports plugin
+	 *
+	 * @return json
+	 */
+	public function post()
+	{
+		$app         = JFactory::getApplication();
+		$jinput      = $app->input;
+		$formData    = $jinput->post;
+		$reportName  = $formData->get('report');
```
Comment (vaivk369) [inline src/reports/report.php]: @monikaatvowel  will this work if I pass the reportname is the url?

File: src/language/en-GB/en-GB.plg_api_reports.ini [hunk 10]
```diff
@@ -0,0 +1,2 @@
+PLG_API_REPORTS_REPORT_NAME_MISSSING="Please enter report name"
```
Comment (manojLondhe) [inline src/language/en-GB/en-GB.plg_api_reports.ini:6]: Add copyright

File: src/language/en-GB/en-GB.plg_api_reports.sys.ini [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
```
Comment (manojLondhe) [inline src/language/en-GB/en-GB.plg_api_reports.sys.ini]: remove spaces before and after `=`

File: src/reports/report.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+		$report = $reportPlugin->getItems();
+
+		$errors = $reportPlugin->getTJRMessages();
+
+		if (!empty($errors))
+		{
+			ApiError::raiseError(400, $errors[0], 'APIValidationException');
+		}
+
+		if (!empty($reportCols))
```
Comment (manojLondhe) [inline src/reports/report.php]: handle else first?

File: src/reports/report.php [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	 * @return json
+	 */
+	public function post()
+	{
+		$app         = JFactory::getApplication();
+		$jinput      = $app->input;
+		$formData    = $jinput->post;
+		$reportName  = $app->input->get('id');
+
+		if (empty($reportName)) {
```
Comment (manojLondhe) [inline src/reports/report.php]: not phpced?

**Review by manojLondhe — CHANGES_REQUESTED** [review]

File: src/language/en-GB/en-GB.plg_api_reports.sys.ini (hunk 2 above)
Comment (manojLondhe) [reply src/language/en-GB/en-GB.plg_api_reports.sys.ini]: use `PLG_API_REPORTS`

File: src/reports/report.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	 * Function to get report data from tjreports plugin
+	 *
+	 * @return json
+	 */
+	public function post()
+	{
+		$app         = JFactory::getApplication();
+		$jinput      = $app->input;
+		$formData    = $jinput->post;
+		$reportName  = $app->input->get('id');
```
Comment (vaivk369) [inline src/reports/report.php]: this should be string right?

File: src/reports.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * @version     CVS: 1.0.0
+ * @package     Joomla.Plugin
+ * @subpackage  Api.reports
```
Comment (thite-amol) [inline src/reports.php]: @monikaatvowel Add standard doc comment's

File: src/reports/filters.php [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * Function get filters data
+	 *
+	 * @return boolean
+	 */
+	public function get()
+	{
+		$app         = JFactory::getApplication();
+		$jinput      = $app->input->get;
+		$reportName  = $jinput->get('report');
```
Comment (thite-amol) [inline src/reports/filters.php]: @monikaatvowel Add filters

File: src/reports/filters.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+		JLoader::import('plugins.tjreports.' . $reportName . "." . $reportName, JPATH_SITE);
+		$classname = 'TjreportsModel' . ucfirst($reportName);
+		$reportPlugin = new $classname;
+
+		$filters = $reportPlugin->displayFilters();
```
Comment (thite-amol) [inline src/reports/filters.php]: @monikaatvowel verify the class is available

File: src/language/en-GB/en-GB.plg_api_reports.ini (hunk 1 above)
Comment (monikaatvowel) [reply src/language/en-GB/en-GB.plg_api_reports.ini]: changed to PLG_API_REPORTS

File: src/reports/filters.php (hunk 17 above)
Comment (monikaatvowel) [reply src/reports/filters.php]: done

File: src/language/en-GB/en-GB.plg_api_reports.sys.ini (hunk 2 above)
Comment (monikaatvowel) [reply src/language/en-GB/en-GB.plg_api_reports.sys.ini]: done

File: src/reports.php (hunk 15 above)
Comment (monikaatvowel) [reply src/reports.php]: updated

File: src/reports.xml (hunk 3 above)
Comment (monikaatvowel) [reply src/reports.xml]: done

File: src/reports/filters.php (hunk 4 above)
Comment (monikaatvowel) [reply src/reports/filters.php]: changed to appropriate comment

File: src/reports/filters.php (hunk 5 above)
Comment (monikaatvowel) [reply src/reports/filters.php]: fixed auth code

File: src/reports/filters.php (hunk 16 above)
Comment (monikaatvowel) [reply src/reports/filters.php]: done

File: src/reports/report.php (hunk 6 above)
Comment (monikaatvowel) [reply src/reports/report.php]: done

File: src/reports/report.php (hunk 9 above)
Comment (monikaatvowel) [reply src/reports/report.php]: changed to handle both types of requests

File: src/reports/report.php (hunk 7 above)
Comment (monikaatvowel) [reply src/reports/report.php]: done

File: src/reports/report.php (hunk 8 above)
Comment (monikaatvowel) [reply src/reports/report.php]: done

File: src/reports/report.php (hunk 12 above)
Comment (monikaatvowel) [reply src/reports/report.php]: added Todo

**Review by manojLondhe — APPROVED** [review]

**Comment by manojLondhe** [comment]
@monikaatvowel Thank you for PR :+1:

**Pull request merged**